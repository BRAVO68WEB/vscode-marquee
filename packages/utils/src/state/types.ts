import { ExtensionContext } from 'vscode'

export abstract class StateProvider<State> {
  abstract context: ExtensionContext
  abstract state: State

  /**
   * Allows to sync a specific key value pair with provider
   * @param key state key
   * @param value state value
   */
  abstract updateState (key: keyof State, value: State[keyof State]): Promise<void>
  /**
   * Downloads the latest state from the provider and returns it
   */
  abstract getState (): State
  /**
   * Gets the provider values and compares them with local ones. If a conflict
   * is detected the value of the environment that was last updated is picked.
   * ToDo(Christian): make this configurable, e.g. allow to get asked etc.
   */
  abstract sync (): State
  /**
   * Clears the state of the provider
   */
  abstract clear (): Promise<void>
}
