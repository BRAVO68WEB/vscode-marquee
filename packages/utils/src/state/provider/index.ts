import type { ExtensionContext } from 'vscode'

import { StatefulState } from './stateful'
import type { StateProvider } from '../types'

type StateProviderClass = new (
  context: ExtensionContext,
  key: string,
  defaultState: any
) => StateProvider<any>

const provider: Record<string, StateProviderClass> = {
  stateful: StatefulState
}

export default provider
