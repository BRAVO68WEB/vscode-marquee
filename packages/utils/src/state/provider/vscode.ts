import pick from 'lodash.pick'
import hash from 'object-hash'
import type { ExtensionContext } from 'vscode'

import { Logger } from '../../logger'
import { DEPRECATED_GLOBAL_STORE_KEY } from '../../constants'
import type { StateProvider } from '../types'

export class VSCodeState<State> implements StateProvider<State> {
  #state: State

  constructor (
    public context: ExtensionContext,
    private _key: string,
    private _defaultState: State
  ) {
    const oldGlobalStore = this.context.globalState.get<object>(DEPRECATED_GLOBAL_STORE_KEY, {})
    this.#state = {
      ...this._defaultState,
      ...pick(oldGlobalStore, Object.keys(this._defaultState as any)),
      ...this.context.globalState.get<State>(this._key)
    }

    /**
     * preserve state across different machines
     */
    this.context.globalState.setKeysForSync(Object.keys(this._defaultState as any))
  }

  get state () {
    return this.#state
  }

  /**
   * Update extension state
   * @param prop state property name
   * @param val new state property value
   * @param broadcastState set to true if you want to broadcast this state change
   *                       (only needed when updating state from the extension host)
   */
  async updateState <T extends keyof State = keyof State>(prop: T, val: State[T]) {
    /**
     * check if we have to update
     */
    if (
      typeof val !== 'undefined' &&
      typeof this.#state[prop] !== 'undefined' &&
      hash(this.#state[prop] as any) === hash(val as any)
    ) {
      return
    }

    Logger.info(`Update state "${prop.toString()}": ${val as any as string}`)
    this.#state[prop] = val
    await this.context.globalState.update(this._key, this.#state)
  }

  async clear () {
    this.#state = { ...this._defaultState }
    await this.context.globalState.update(this._key, this.#state)
    await this.context.globalState.update(DEPRECATED_GLOBAL_STORE_KEY, undefined)
  }

  getState (): State {
    throw new Error('Method not implemented.')
  }

  sync (): State {
    return this.state
  }
}
