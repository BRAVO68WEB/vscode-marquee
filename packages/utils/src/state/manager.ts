import { Disposable, ExtensionContext } from 'vscode'

import provider from './provider'
import { VSCodeState } from './provider/vscode'
import { Logger } from '../logger'
import type { StateProvider } from './types'

export class StateManager<State> implements Disposable {
  #state?: State
  #logger: Logger
  #context: ExtensionContext
  #provider: StateProvider<State>[] = []

  constructor (context: ExtensionContext, key: string, defaultState: State) {
    this.#logger = Logger.getChildLogger('StateManager')
    this.#context = context
    this.#provider.push(
      /**
       * always
       */
      new VSCodeState(this.#context, key, defaultState),
      ...(this.#context.workspaceState.get<string[]>('stateProvider') || [])
        .filter((p) => Boolean(provider[p]))
        .map((p) => new provider[p](context, key, defaultState))
    )
  }

  get state (): State  {
    if (!this.state) {
      throw new Error('StateManager is not in sync')
    }

    return this.state
  }

  async updateState <T extends keyof State = keyof State>(prop: T, val: State[T]) {
    for (const provider of this.#provider) {
      await provider.updateState(prop, val)
    }
  }

  async clearState () {
    for (const provider of this.#provider) {
      await provider.clear()
    }
  }

  /**
   * Downloads current State from providers and merge them if necessary
   */
  async sync () {
    for (const provider of this.#provider) {
      await provider.sync()
    }
  }

  dispose () {
    // nothing to do
  }
}
