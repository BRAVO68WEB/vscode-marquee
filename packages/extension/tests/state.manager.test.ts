import vscode from 'vscode';
import { StateManager } from '../src/state.manager';

const context = {} as any;

describe('StateManager', () => {
  test('updateWorkspaces', () => {
    jest.spyOn(StateManager.prototype, 'handlePersistence')
      .mockImplementation(jest.fn());
    jest.spyOn(StateManager.prototype, 'getActiveWorkspace')
      .mockImplementation(jest.fn());
    const state = new StateManager(context as any);
    expect(state.updateWorkspaces({})).toMatchSnapshot();
    expect(state.updateWorkspaces({ activeWorkspaceId: 'foobar' }))
      .toMatchSnapshot();

    // @ts-expect-error mock
    state.getActiveWorkspace.mockReturnValue({ id: 'foo' });
    expect(state.updateWorkspaces({})).toMatchSnapshot();
    expect(state.updateWorkspaces({ workspaces: [{ id: 'foo' }] }))
      .toMatchSnapshot();

    // @ts-expect-error mock
    vscode.env.remoteName = 'codespaces';
    expect(state.updateWorkspaces({ workspaces: [{ id: 'foobar' }] }))
      .toMatchSnapshot();
  });
});
