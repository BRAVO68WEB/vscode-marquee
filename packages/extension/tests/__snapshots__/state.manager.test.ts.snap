// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateManager updateWorkspaces 1`] = `
Object {
  "workspaces": Array [],
}
`;

exports[`StateManager updateWorkspaces 2`] = `
Object {
  "activeWorkspaceId": undefined,
  "workspaces": Array [],
}
`;

exports[`StateManager updateWorkspaces 3`] = `
Object {
  "workspaces": Array [
    Object {
      "id": "foo",
    },
  ],
}
`;

exports[`StateManager updateWorkspaces 4`] = `
Object {
  "workspaces": Array [
    Object {
      "id": "foo",
    },
  ],
}
`;

exports[`StateManager updateWorkspaces 5`] = `
Object {
  "workspaces": Array [
    Object {
      "id": "foobar",
    },
  ],
}
`;
